EIPS/eip-2537.md

```diff
@@ -0,0 +1,293 @@
+
+Scalar for multiplication operation is encoded as `32` bytes by performing BigEndian encoding of the corresponding (unsigned) integer. Corresponding integer is **not** required to be less than or equal than main subgroup size.
+
+#### ABI for operations
+
+##### ABI for G1 addition
+
+G1 addition call expects `256` bytes as an input that is interpreted as byte concatenation of two G1 points (`128` bytes each). Output is an encoding of addition operation result - single G1 point (`128` bytes).
+
+Error cases:
+- Either of points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G1 multiplication
+
+G1 multiplication call expects `160` bytes as an input that is interpreted as byte concatenation of encoding of G1 point (`128` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiplication operation result - single G1 point (`128` bytes).
+
+Error cases:
+- Point being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G1 multiexponentiation
+
+G1 multiplication call expects `160*k` bytes as an input that is interpreted as byte concatenation of `k` slices each of them being a byte concatenation of encoding of G1 point (`128` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiexponentiation operation result - single G1 point (`128` bytes).
+
+Error cases:
+- Any of G1 points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 addition
+
+G2 addition call expects `512` bytes as an input that is interpreted as byte concatenation of two G2 points (`256` bytes each). Output is an encoding of addition operation result - single G2 point (`256` bytes).
+
+Error cases:
+- Either of points being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 multiplication
+
+G2 multiplication call expects `288` bytes as an input that is interpreted as byte concatenation of encoding of G2 point (`256` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiplication operation result - single G2 point (`256` bytes).
+
+Error cases:
+- Point being not on the curve must result in error
+- Field elements encoding rules apply (obviously)
+- Input has invalid length
+
+##### ABI for G2 multiexponentiation
+
+G2 multiplication call expects `288*k` bytes as an input that is interpreted as byte concatenation of `k` slices each of them being a byte concatenation of encoding of G2 point (`256` bytes) and encoding of a scalar value (`32` bytes). Output is an encoding of multiexponentiation operation result - single G2 point (`256` bytes).
+
+Error cases:
+- Any of G2 points being not on the curve must result in error
```

Comment:
What is the gas charge in this case? All gas sent or gas used up until that point?

Reply:
I'm not familiar with current procedure of what happens if the current BN precompile fails or any other precompile fails. I'd advise that it must consume all the gas, but it's out of scope for this spec and it's EVM internals.

Reply:
I think it is in scope for this EIP. It has to clearly describe how the precompile behaves.

Reply:
Well, so far I do not see any mechanism how precompile can affect the VM. At the moment precompiles take input from memory and return `(bytes memory result, bool success)` (pseudo-signature of `CALL` opcode and it's relatives), so no matter what will be an error inside and whether it's on the last execution step or the first it'll consume some gas (what is a behavior of the current precompiles?) and return `success = false`. I've added a chapter that is `MUST` consume an amount of gas from the gas schedule, but I don't know how it's implemented in EVM now.

Reply:
The CALL frame to the precompile has its own gas limit and once the precompile returns control it has to state the total gas used and the execution state (success or failure). This specification has to be clear when and what those returning values are.

Reply:
The spec is clear about what are return values and error cases, as well as the gas costs. But how those are deduced by EVM and other glue is not in my power to change.

Reply:
> The spec is clear about what are return values and error cases, as well as the gas costs. But how those are deduced by EVM and other glue is not in my power to change.

Sorry I am not sure I understand what you mean. Could you explain?

Reply:
This spec is clear about what gas caller should pay to execute the precompile and when precompile will return an error. Yes, logically in error case a full amount of gas should be burned, but as you mentioned it’s not a precompile machinery, but some other EVM internals which this spec does not touch.

Reply:
Specifying what happens in the error case is definitely the task of the precompile. Saying "all supplied gas is consumed in case of an error" would be enough.

Reply:
It’s already in a spec